### PR TITLE
Add flow: initial types

### PR DESCRIPTION
**Pull request by hzoo**
flow quick reference: http://flowtype.org/docs/quick-reference.html

Ref https://github.com/cst/cst/issues/42

I did `npm i flow-bin -g` and ran `flow check` (flow@0.18.1)

Can potentially remove some of the jsdoc stuff? `* @returns {String}`

### CODE REVIEW

**Comment by mdevils** [comment]
Interesting idea. Will take a closer look later.

File: .flowconfig [hunk 1]
```diff
@@ -0,0 +1,10 @@
+[ignore]
+.*/node_modules/.*
+
+[include]
+
+[libs]
+interfaces/
+
+[options]
+unsafe.enable_getters_and_setters=true
```
Comment (hzoo) [inline .flowconfig]: getters/setters aren't support that well but I found this option.

File: src/Parser.js [hunk 2]
```diff
@@ -1,13 +1,17 @@
+/* @flow */
+
 import {parse} from 'babel-core';
 
 import {buildTokenList, buildElementTree} from './elementTree';
 
 export default class Parser {
+    _strictModeEnabled: boolean;
```
Comment (hzoo) [inline src/Parser.js]: this defines types outside of class (`this. _strictModeEnabled`)

**Comment by hzoo** [comment]
The types here aren't that specific for a lot of stuff - but we can define our own types, etc. Stuff like `Array<any>` could be more specific.

File: src/elements/Element.js [hunk 3]
```diff
@@ -753,11 +766,11 @@ export default class Element {
...
      * @returns {Element}
      */
-    cloneElement() {
+    cloneElement(): Element {
         let clonedChildren = new Array(this._childElements.length);
         for (let i = 0; i < clonedChildren.length; i++) {
             clonedChildren[i] = this._childElements[i].cloneElement();
         }
-        return new this.constructor(clonedChildren);
+        return new this.constructor(this._type, clonedChildren);
```
Comment (hzoo) [inline src/elements/Element.js]: found an issue with `cloneElement` through flow (I think)?

File: src/elements/Element.js [hunk 4]
```diff
@@ -21,7 +24,7 @@ export default class Element {
         this._parentElement = null;
         this._nextSibling = null;
         this._previousSibling = null;
-        this._childElements = null;
+        this._childElements = [];
```
Comment (hzoo) [inline src/elements/Element.js]: is null or [] a better initial value?

Comment (mdevils) [inline src/elements/Element.js]: Null was used to avoid spending resources on unused array I think.

File: src/elements/Element.js [hunk 5]
```diff
@@ -33,13 +36,24 @@ export default class Element {
...
+    isModuleSpecifier: ?string;
+
     /**
      * Element type.
      * Important: some Node types also exist in Tokens. Do not rely on type only, use isToken, isNode.
      *
      * @returns {String}
      */
-    get type() {
+    get type(): string {
```
Comment (mdevils) [inline src/elements/Element.js]: Do they only use lower case? i.e. `string`, `boolean`? Can't we use same names as in ES like `String`, `Boolean`?

Comment (hzoo) [inline src/elements/Element.js]: There is primitive and class types - http://flowtype.org/docs/quick-reference.html#built-in-primitive-types but I think it's different

File: src/elements/Element.js (hunk 4 above)
Comment (hzoo) [inline src/elements/Element.js]: Ok I just need to figure out the syntax to allow that

**Comment by hzoo** [comment]
Still have errors but fixed merge conflicts

File: interfaces/babylon.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
+type babylonOptions = {
```
Comment (mdevils) [inline interfaces/babylon.js]: Can you please add a comment to this file to describe its purpose?

File: src/Parser.js [hunk 7]
```diff
@@ -80,17 +81,19 @@ export default class Parser {
...
-    getOptions() {
+    getOptions(): Object {
         return this._options;
     }
 
     /**
      * @param {CSTParserOptions} newOptions
      */
-    setOptions(newOptions) {
+    setOptions(newOptions: Object) {
```
Comment (mdevils) [inline src/Parser.js]: I think we can introduce type for these options too.

File: src/Parser.js [hunk 8]
```diff
@@ -110,13 +113,13 @@ export default class Parser {
         };
     }
 
-    parse(code) {
+    parse(code: string): Object {
```
Comment (mdevils) [inline src/Parser.js]: This method should return `Program`

File: src/elementTree.js [hunk 9]
```diff
@@ -1,7 +1,22 @@
...
+
+import type Program from './elements/types/Program';
+import type Location from './elements/Element';
+
 import visitorKeys from './visitorKeys';
 import elementIndex from './elements/elementIndex';
 import Token from './elements/Token';
-import * as babylon from 'babylon';
+
+type BabylonToken = {
```
Comment (mdevils) [inline src/elementTree.js]: I think we can move this interface to `interfaces/babylon`?

**Comment by mdevils** [comment]
Good job! Added some notes here.

**Comment by hzoo** [comment]
Updated. We could also move types into separate files depending on how you want to do it. Also more examples of types - https://github.com/facebook/flow/tree/master/lib - maybe check out the react one too

**Comment by mdevils** [comment]
Great job!

**Pull request merged**